File: backend/app/schemas/attendance.py
"""Pydantic v2 DTOs — Attendance Engine."""
from __future__ import annotations
from datetime import date, datetime, time
from typing import List, Optional
from pydantic import BaseModel, Field

from app.models.attendance import AttendanceStatusEnum, DeviceActionEnum


class CheckInMetaData(BaseModel):
    latitude: float = Field(..., ge=-90.0, le=90.0)
    longitude: float = Field(..., ge=-180.0, le=180.0)
    gps_accuracy: float = Field(..., ge=0.0)
    device_id: Optional[str] = None
    device_model: Optional[str] = None
    os_version: Optional[str] = None
    app_version: Optional[str] = None
    connection_type: Optional[str] = None
    face_embedding: Optional[bytes] = None  # Nullable BLOB for future face recognition


class CheckOutMetaData(BaseModel):
    latitude: float = Field(..., ge=-90.0, le=90.0)
    longitude: float = Field(..., ge=-180.0, le=180.0)
    gps_accuracy: float = Field(..., ge=0.0)
    device_id: Optional[str] = None
    device_model: Optional[str] = None
    os_version: Optional[str] = None
    app_version: Optional[str] = None
    connection_type: Optional[str] = None
    task_summary: Optional[str] = Field(None, max_length=2000)


class AttendanceDeviceLogResponse(BaseModel):
    id: int
    action_type: DeviceActionEnum
    device_id: Optional[str]
    device_model: Optional[str]
    os_version: Optional[str]
    app_version: Optional[str]
    ip_address: Optional[str]
    connection_type: Optional[str]
    latitude: Optional[float]
    longitude: Optional[float]
    gps_accuracy: Optional[float]
    photo_path: Optional[str]
    captured_at: datetime

    model_config = {"from_attributes": True}


class EmployeeMinimal(BaseModel):
    id: int
    employee_code: str
    first_name: str
    last_name: str
    full_name: str
    email: str
    photo_path: Optional[str] = None

    model_config = {"from_attributes": True}


class AttendanceResponse(BaseModel):
    id: int
    employee_id: int
    employee: Optional[EmployeeMinimal] = None
    date: date
    check_in_time: Optional[time] = None
    check_in_photo_path: Optional[str] = None
    check_in_latitude: Optional[float] = None
    check_in_longitude: Optional[float] = None
    check_in_address: Optional[str] = None
    check_in_google_maps_url: Optional[str] = None
    check_in_gps_accuracy: Optional[float] = None

    check_out_time: Optional[time] = None
    check_out_photo_path: Optional[str] = None
    check_out_latitude: Optional[float] = None
    check_out_longitude: Optional[float] = None
    check_out_address: Optional[str] = None
    check_out_google_maps_url: Optional[str] = None
    check_out_gps_accuracy: Optional[float] = None
    checkout_task_summary: Optional[str] = None

    working_hours: Optional[float] = None
    status: AttendanceStatusEnum
    remarks: Optional[str] = None
    is_regularized: bool
    device_logs: List[AttendanceDeviceLogResponse] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AttendanceTodaySummary(BaseModel):
    total_employees: int = 0
    present_today: int = 0
    absent_today: int = 0
    late_today: int = 0
    on_leave_today: int = 0
    wfh_today: int = 0
    on_duty_today: int = 0


class RegularizeAttendanceRequest(BaseModel):
    status: AttendanceStatusEnum
    check_in_time: Optional[time] = None
    check_out_time: Optional[time] = None
    remarks: str = Field(..., min_length=3)

File: admin-panel/src/api/attendance.ts
import apiClient from "./client";

export interface AttendanceItem {
  id: number;
  employee_id: number;
  employee?: {
    id: number;
    employee_code: string;
    first_name: string;
    last_name: string;
    full_name: string;
    email: string;
    photo_path?: string;
  };
  date: string;
  check_in_time?: string;
  check_in_photo_path?: string;
  check_in_latitude?: number;
  check_in_longitude?: number;
  check_in_address?: string;
  check_in_google_maps_url?: string;
  check_in_gps_accuracy?: number;

  check_out_time?: string;
  check_out_photo_path?: string;
  check_out_latitude?: number;
  check_out_longitude?: number;
  check_out_address?: string;
  check_out_google_maps_url?: string;
  check_out_gps_accuracy?: number;
  checkout_task_summary?: string;

  working_hours?: number;
  status: string;
  remarks?: string;
  is_regularized: boolean;
  device_logs?: Array<{
    id: number;
    action_type: string;
    device_id?: string;
    device_model?: string;
    os_version?: string;
    app_version?: string;
    ip_address?: string;
    connection_type?: string;
    latitude?: number;
    longitude?: number;
    gps_accuracy?: number;
    photo_path?: string;
    captured_at: string;
  }>;
  created_at: string;
  updated_at: string;
}

export interface AttendanceTodaySummaryData {
  total_employees: number;
  present_today: number;
  absent_today: number;
  late_today: number;
  on_leave_today: number;
  wfh_today: number;
  on_duty_today: number;
}

export const fetchAttendanceList = async (params: {
  company_id: number;
  page?: number;
  page_size?: number;
  department_id?: number;
  status?: string;
  from_date?: string;
  to_date?: string;
  search?: string;
}) => {
  const response = await apiClient.get("/attendance", { params });
  return response.data;
};

export const fetchTodaySummary = async (company_id: number = 1): Promise<AttendanceTodaySummaryData> => {
  const response = await apiClient.get("/attendance/today-summary", { params: { company_id } });
  return response.data;
};

export const fetchAttendanceDetail = async (id: number): Promise<AttendanceItem> => {
  const response = await apiClient.get(`/attendance/${id}`);
  return response.data;
};

export const regularizeAttendance = async (
  id: number,
  payload: { status: string; remarks: string; check_in_time?: string; check_out_time?: string }
) => {
  const response = await apiClient.post(`/attendance/${id}/regularize`, payload);
  return response.data;
};
